### PR TITLE
[2-36] HR Assign Test Page Creation

### PR DESCRIPTION
**Pull request by KuznetsovR**

### CODE REVIEW

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+			<mat-label>Expire date</mat-label>
+			<input matInput [min]="minDate" [matDatepicker]="picker" (dateChange)="onDateChange($event)" />
+			<mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
+			<mat-datepicker #picker></mat-datepicker>
+		</mat-form-field>
+	</div>
+	<div class="link-container">generated link will be here</div>
+	<mat-checkbox class="email-checkbox" [checked]="sendEmail" (change)="toggle($event)"
+		>Send email to user</mat-checkbox
+	>
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html]: Please, cope with this alone bracket

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<h1 mat-dialog-title>Assign test on {{data.name}}</h1>
+<div mat-dialog-content>
+	<div>
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html]: Please don't use empty tags without reasons

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+		<mat-form-field appearance="fill" class="modal-parameter">
+			<mat-label>English level</mat-label>
+			<mat-select (selectionChange)="onLevelChange($event)">
+				<mat-option *ngFor="let level of levels" [value]="level"> {{level}} </mat-option>
+			</mat-select>
+		</mat-form-field>
+	</div>
+	<div>
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html]: Same as above

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	.mat-dialog-container {
+		text-align: center;
+	}
+	.mat-dialog-content {
+		overflow: visible;
+	}
+}
+.assign-btn {
+	justify-content: center;
+	font-size: 1.2rem;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss]: Maybe we can use our font variable here? Or create new for buttons

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	line-height: 1.2rem;
+}
+.link-container {
+	background: $light-gray;
+	font-size: 1.2rem;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss]: Same here

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss (hunk 4 above)
Comment (YanaBr) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss]: Also why you use rem here? Is it intentional?

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import { MatSelectChange } from '@angular/material/select';
+import { MatCheckboxChange } from '@angular/material/checkbox';
+
+@Component({
+	selector: 'app-assign-test-modal',
+	templateUrl: './assign-test-modal.component.html',
+	styleUrls: ['./assign-test-modal.component.scss']
+})
+export class AssignTestModalComponent {
+	level: string | null;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts:14]: Good practice: separate English level to enum

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	date: Date | null;
+	sendEmail = true;
+	minDate: Date;
+	maxDate: Date;
+	constructor(
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts:22]: Space is missed between constructor and variables

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+		public dialogRef: MatDialogRef<AssignTestModalComponent>,
+		@Inject(MAT_DIALOG_DATA) public data: { name: string }
+	) {
+		const currentDate = new Date();
+		this.minDate = new Date(currentDate.valueOf() + 86400000);
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts]: Please, explain this magic number (maybe leave a comment)

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	onLevelChange(event: MatSelectChange): void {
+		this.level = event.value;
+	}
+	onDateChange(event: MatDatepickerInputEvent<Date>): void {
+		this.date = event.value;
+	}
+	toggle(event: MatCheckboxChange): void {
+		this.sendEmail = event.checked;
+	}
+	levels: string[] = ['A1', 'A2', 'B1', 'B2', 'C1', 'C2'];
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts]: As I told above you can move it to enum

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html [hunk 10]
```diff
@@ -0,0 +1,31 @@
+<div class="btn-wrapper">
+	<button mat-raised-button color="primary" class="go-back-btn"><a (click)="goBack()">BACK</a></button>
+</div>
+<div class="history-text"><b> TESTS HISTORY</b></div>
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html:4]: Is it an extra space?

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+<form class="example-form">
+	<mat-form-field class="example-full-width" appearance="fill">
+		<mat-label>search</mat-label>
+		<input matInput placeholder="Enter name" (input)="applyFilter($event)" />
+	</mat-form-field>
+</form>
+
+<div class="mat-elevation-z8 user-table">
+	<table mat-table [dataSource]="dataSource" matSort style="width: 100%; text-align: center">
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html]: Do need inline styles here?

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+		this.location.back();
+	}
+	openDialog(el: { name: string }): void {
+		this.dialog.open(AssignTestModalComponent, {
+			width: '50%',
+			data: el
+		});
+	}
+
+	constructor(private tableService: TableService, private location: Location, public dialog: MatDialog) {}
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts]: Please move constructor to the top of the class

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+	ngAfterViewInit(): void {
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts:42]: Same as above

File: exalearn-ui/src/app/services/table.service.ts [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import { Injectable } from '@angular/core';
```
Comment (YanaBr) [inline exalearn-ui/src/app/services/table.service.ts:2]: Please leave TODO to remove it in future

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts (hunk 8 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts]: It's 24 hours in miiliseconds, so hr couldn't set expiration date on previous days or the current day

File: exalearn-ui/src/app/services/table.service.ts (hunk 14 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/services/table.service.ts:2]: Do you mean the whole service?

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html (hunk 2 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.html]: It's needed as a wrapper with display block, so there is a reason

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss (hunk 4 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.scss]: Created $font-size-button: 1.2rem; yes, I usually use rem

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts (hunk 7 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts:22]: prettier told me not to, but i added space

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts (hunk 9 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts:37]: I tried, but esLint error occured and I couldn't solve it

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html (hunk 10 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html:4]: removed

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html (hunk 11 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.html]: moved to class

File: exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts (hunk 13 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/pages/assign-test-page/assign-test-page.component.ts:42]: It is the same constructor

**Review by KotdenGit — APPROVED** [review]

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts (hunk 9 above)
Comment (KuznetsovR) [reply exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts]: enum added

File: exalearn-ui/src/app/services/table.service.ts (hunk 14 above)
Comment (YanaBr) [reply exalearn-ui/src/app/services/table.service.ts:2]: Now here us only method with test data, so yes, just keave a comment TODO to remove test sata after backed integration

File: exalearn-ui/src/app/Enums/enums.ts [hunk 15]
```diff
@@ -0,0 +1,8 @@
+export enum EnglishLevels {
+	A1 = 'Beginner(A1)',
+	A2 = 'Elementary(A2)',
+	B1 = 'Intermediate(B1)',
+	B2 = 'Upper-Intermediate(A2)',
+	C1 = 'Advanced(C1)',
+	C2 = 'Proficiency(C2)'
+}
```
Comment (vstrabko) [inline exalearn-ui/src/app/Enums/enums.ts:8]: Just a suggestion. Maybe space before brackets?
If it looks good on the page, than it's OK.

**Review by YanaBr — APPROVED** [review]

File: exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts [hunk 16]
```diff
@@ -15,6 +16,8 @@ export class AssignTestModalComponent {
 	sendEmail = true;
 	minDate: Date;
 	maxDate: Date;
+	levels = EnglishLevels;
+	levelsValues = Object.values(this.levels);
```
Comment (vstrabko) [inline exalearn-ui/src/app/components/assign-test-modal/assign-test-modal.component.ts:21]: Try to get rid of this line. It is not needed.

Consider creating a custom pipe or a better method to read directly from ENUM   :)

**Review by vstrabko — CHANGES_REQUESTED** [review]

**Comment by vstrabko** [comment]
okay(

**Pull request merged**